src/main/java/com/billyyccc/mssqlclient/impl/codec/InitCommandCodec.java

```diff
@@ -0,0 +1,227 @@
+import com.billyyccc.mssqlclient.impl.utils.Utils;
+import io.vertx.sqlclient.impl.Connection;
+import io.vertx.sqlclient.impl.command.InitCommand;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import static java.nio.charset.StandardCharsets.UTF_16LE;
+
+class InitCommandCodec extends MSSQLCommandCodec<Connection, InitCommand> {
+  InitCommandCodec(InitCommand cmd) {
+    super(cmd);
+  }
+
+  @Override
+  void encode(TdsMessageEncoder encoder) {
+    super.encode(encoder);
+    sendLoginMessage();
+  }
+
+  @Override
+  void decodeMessage(TdsMessage message, TdsMessageEncoder encoder) {
+    ByteBuf messageBody = message.content();
+    while (messageBody.isReadable()) {
+      byte tokenType = messageBody.readByte();
+      switch (tokenType) {
+        //FIXME complete all the logic here
+        case DataPacketStreamTokenType.LOGINACK_TOKEN:
+          result = cmd.connection();
+          break;
+        case DataPacketStreamTokenType.ERROR_TOKEN:
+          handleErrorToken(messageBody);
+          break;
+        case DataPacketStreamTokenType.INFO_TOKEN:
+          break;
+        case DataPacketStreamTokenType.ENVCHANGE_TOKEN:
+          break;
+        case DataPacketStreamTokenType.DONE_TOKEN:
+          handleDoneToken();
+          break;
+      }
+    }
+  }
+
+  private void sendLoginMessage() {
+    ChannelHandlerContext chctx = encoder.chctx;
+
+    ByteBuf packet = chctx.alloc().ioBuffer();
+
+    // packet header
+    packet.writeByte(MessageType.TDS7_LOGIN.value());
+    packet.writeByte(MessageStatus.NORMAL.value() | MessageStatus.END_OF_MESSAGE.value());
+    int packetLenIdx = packet.writerIndex();
+    packet.writeShort(0); // set length later
+    packet.writeShort(0x00);
+    packet.writeByte(0x00); // FIXME packet ID
+    packet.writeByte(0x00);
+
+    int startIdx = packet.writerIndex(); // Length
+    packet.writeInt(0x00); // set length later by calculating
+    packet.writeInt(LoginPacket.SQL_SERVER_2017_VERSION); // TDSVersion
+    packet.writeIntLE(LoginPacket.DEFAULT_PACKET_SIZE); // PacketSize
+    packet.writeIntLE(0x00); // ClientProgVer
+    packet.writeIntLE(0x00); // ClientPID
+    packet.writeIntLE(0x00); // ConnectionID
+    packet.writeByte(LoginPacket.DEFAULT_OPTION_FLAGS1
+      | LoginPacket.OPTION_FLAGS1_DUMPLOAD_OFF); // OptionFlags1
+    packet.writeByte(LoginPacket.DEFAULT_OPTION_FLAGS2); // OptionFlags2
+    packet.writeByte(LoginPacket.DEFAULT_TYPE_FLAGS); // TypeFlags
+    packet.writeByte(LoginPacket.DEFAULT_OPTION_FLAGS3); // OptionFlags3
+    packet.writeIntLE(0x00); // ClientTimeZone
+    packet.writeIntLE(0x00); // ClientLCID
+
+      /*
+        OffsetLength part:
+        we use a map to store offset related information:
+        parameter name -> packet ByteBuf writer position for offset
+        then we set offset by calculating ByteBuf writer indexes diff later
+       */
+    Map<String, Integer> offsetInfo = new HashMap<>();
```

Comment:
this map does not seem necessary and can be replaced with local variables which are more efficient

Reply:
updated with https://github.com/BillyYccc/incubator-vertx-mssql-client/pull/3/commits/807cc56bd08d4a7a703e805cebf60e5ff0193b54.